File: project_euler/Problem046/Problem046.py
import math

class Problem046:
    def solution(self) -> None:
        primes, result = {2}, 3
        while True:
            if self.__is_prime(result):
                primes.add(result)
            else:
                border, is_goldbach = int(math.sqrt((result - 1) >> 1)), False
                for i in range(1, border + 1):
                    if result - (i * i << 1) in primes:
                        is_goldbach = True
                        break
                if not is_goldbach:
                    print(result)
                    return
            result += 2

    def __is_prime(self, value: int) -> bool:
        if 1 != value % 6 and 5 != value % 6:
            return 2 == value or 3 == value

        border = int(math.sqrt(value))
        for i in range(5, border + 1, 6):
            if not value % i or not value % (i + 2):
                return False
        return 1 != value

File: project_euler/Problem046/Problem046.java
package project_euler;

import java.util.HashSet;
import java.util.Set;

public class Problem046 {
    public void solution() {
        Set<Integer> primes = new HashSet<>();
        primes.add(2);
        int result = 3;
        while (true) {
            if (isPrime(result)) {
                primes.add(result);
            } else {
                int border = (int) Math.sqrt((result - 1) >> 1);
                boolean isGoldbach = false;
                for (int i = 1; i <= border; ++i) {
                    if (primes.contains(result - (i * i << 1))) {
                        isGoldbach = true;
                        break;
                    }
                }
                if (!isGoldbach) {
                    System.out.println(result);
                    return;
                }
            }
            result += 2;
        }
    }

    private boolean isPrime(int value) {
        if (1 != value % 6 && 5 != value % 6) {
            return 2 == value || 3 == value;
        }

        int border = (int) Math.sqrt(value);
        for (int i = 5; i <= border; i += 6) {
            if (0 == value % i || 0 == value % (i + 2)) {
                return false;
            }
        }
        return 1 != value;
    }
}
